README.md
rider
===

[![NPM version](https://img.shields.io/npm/v/rider.svg?style=flat-square)](https://npmjs.org/package/rider) [![Build Status](https://img.shields.io/travis/ecomfe/rider.svg?style=flat-square)](https://travis-ci.org/ecomfe/rider) [![License](https://img.shields.io/npm/l/rider.svg?style=flat-square)](./LICENSE) [![EFE Mobile Team](https://img.shields.io/badge/EFE-Mobile_Team-blue.svg?style=flat-square)](http://efe.baidu.com)

`rider` 是基于 [Stylus](http://stylus-lang.com) 与后处理器、无侵入风格的 CSS 样式工具库。

基于 `rider` 开发 **项目** 或 **UI 样式库**，能有效提高开发效率，延长代码生命周期。

## 功能

 * [特性](./doc/feature.md)
 * [样式初始化](./doc/initialize.md)
 * [顺时针简写](./doc/clockhand.md)
 * [缓动函数](./doc/easing.md)
 * [图片](./doc/image.md)
 * [排版](./doc/typography.md)
 * [形状](./doc/shape.md)
 * [动画](./doc/animate.md)
 * [单位转换](./doc/unit.md)
 * [响应式工具](./doc/breakpoint.md)
 * [Retina 支持](./doc/retina.md)
 * [辅助函数](./doc/helper.md)

## 使用

 * 安装与配置
     * [edp 环境](https://github.com/ecomfe/edp-provider-rider)
     * [connect/express 环境](./doc/connect.md)
 * [代码风格](./doc/code-style.md)
 * [常见问题](./doc/faq.md)

## License

MIT &copy; [Baidu Inc.](./LICENSE)


History.md

2.0.0 / 2015-06-17
==================

 * **Breaking Change** 改进 Function：`bg()`, `ir()` [#18](https://github.com/ecomfe/rider/issues/18)
 * **Breaking Change** 改进 Root Mixin：`font-face()` [#16](https://github.com/ecomfe/rider/issues/16) [#24](https://github.com/ecomfe/rider/issues/24)
 * **Breaking Change** 改进 Block Mixin：`+breakpoint()`，移除组合能力，迁出 retina 支持 [#19](https://github.com/ecomfe/rider/issues/19)
 * **Breaking Change** 移除：retina 支持中的浏览器兼容代码 [#17](https://github.com/ecomfe/rider/issues/17)
 * **Breaking Change** 移除 全局变量：`$-image-dppx`，retina 图片机制已更新
 * 增加 全局变量：`$-code-font-family` 作为代码段默认字体
 * 增加 全局变量：`$-image-dppx-range` 作为 retina 图片的探测范围
 * 增加 全局变量：`$-image-dppx-separator` 作为 retina 图片的分隔符
 * 增加 全局变量：`$-image-dppx-quality-first` 作为 retina 图片质量优先模式的开关
 * 增加 全局变量：`$-unit-precision` 作为单位转换的小数点精度控制
 * 增加 Function：`bg2x()`, `bg3x()`, `bg-dppx()` [#23](https://github.com/ecomfe/rider/issues/23)
 * 增加 Function：`file-exists()`, `is-url()`, `get-retina-src()` [#15](https://github.com/ecomfe/rider/issues/15)
 * 增加 Mixin：`normalize()` 作为布局初始化的补充 [#8](https://github.com/ecomfe/rider/issues/8)
 * 增加 Mixin：`circle()` 生成圆形或椭圆
 * 增加 Block Mixin：`+retina()` [#20](https://github.com/ecomfe/rider/issues/20)
 * 改进 Function：`rem()`, `em()`, `px()` 单位转换定义截取小数点位数 [#4](https://github.com/ecomfe/rider/issues/4)
 * 改进 Mixin：`animate()` 升级至 animate.css 3.2.6 版 [#5](https://github.com/ecomfe/rider/issues/5)
 * 改进：测试脚本启用 inline diff [#7](https://github.com/ecomfe/rider/issues/7)
 * 改进：简化 demo 的支持部分 [#6](https://github.com/ecomfe/rider/issues/6)
 * 改进：文档重新经过深入整理 [#21](https://github.com/ecomfe/rider/issues/21)

0.4.4 / 2014-11-06
==================

 * **修复** breakpoint 对最小值的判断
 * **更新** `stylus` 版本至 `0.49.3`（上一个版本存在较严重 bug）

0.4.3 / 2014-09-24
==================

 * 升级开发依赖并测试，将 `stylus` 升级至 `0.49.0`（性能略有提升）

0.4.2 / 2014-08-05
==================

 * **增加** `nav` 元素的 `tap-highlight` 修复
 * 升级开发依赖并测试，其中将 `stylus` 升级至 `0.47.3`

0.4.0 / 2014-06-13
==================

 * **修复** breakpoint 对切片边界值的处理
 * **移除** 对 node 0.8.x 的支持
 * 升级开发依赖并测试，其中将 `stylus` 升级至 `0.46.3`

0.3.4 / 2014-05-12
==================

 * **增加** 动画 `slideInUp`，`slideOutDown`
 * 升级开发依赖并测试，其中将 `stylus` 升级至 `0.45.0`

0.3.3 / 2014-04-28
==================

 * **修复** `animate` 内部实现中的一个逻辑错误
 * **修复** 文档链接
 * 升级开发依赖并测试，其中将 `stylus` 升级至 `0.44.0`

0.3.2 / 2014-04-18
==================

 * 升级开发依赖并测试，将 `stylus` 升级至 `0.43.1`

0.3.1 / 2014-03-25
==================

 * **修改** `initialize()`：增加了表单部分的初始化
 * **修改** `ir()`：删除了冗余的逻辑
 * 完善 `breakpoint` 的测试用例
 * 增加 API、代码风格、FAQ 文档

0.3.0 / 2014-03-17
==================

* 初始版本


test/test.js
/**
 * @file rider test
 * @author firede(firede@firede.us)
 */

var stylus = require('stylus');
var fs = require('fs');
var path = require('path');
var rider = require('../lib/rider');
var assert = require('assert');

var casesPath = path.join(__dirname, './cases');

// test cases
var cases = fs.readdirSync(casesPath).filter(function (file) {
    return ~file.indexOf('.styl');
}).map(function (file) {
    return file.replace('.styl', '');
});


describe('rider test', function () {
    cases.forEach(function (test) {
        var name = test.replace(/[-.]/g, ' ');

        it(name, function () {
            var stylPath = path.join(__dirname, 'cases', test + '.styl');
            var cssPath = path.join(__dirname, 'cases', test + '.css');
            var styl = fs.readFileSync(stylPath, 'utf8').replace(/\r/g, '');
            var css = fs.readFileSync(cssPath, 'utf8').replace(/\r/g, '').trim();

            var style = stylus(styl)
                    .set('filename', stylPath)
                    .set('resolve url', true)
                    .use(rider());

            style.render(function (err, actual) {
                if (err) {
                    throw err;
                }
                assert.equal(actual.trim(), css);
            });
        });
    });
});


demo.js
/**
 * @file rider demo
 * @author firede(firede@firede.us)
 */

var path = require('path');
var http = require('http');

var connect = require('connect');
var serveStatic = require('serve-static');

var stylus = require('stylus');
var rider = require('./lib/rider');
var ap = require('autoprefixer-core');

var app = connect();
var port = 8888;

function compile(str, filePath) {
    return stylus(str)
        .set('filename', filePath)
        .define('url', stylus.resolver())
        .use(rider())
        .on('end', prefixer);
}

function prefixer(_, css) {
    return ap(
        {
            browsers: [
                '> 5%',
                'iOS >= 6',
                'Android >= 2.3',
                'ExplorerMobile >= 10'
            ]
        }
    ).process(css).css;
}

app.use(stylus.middleware({
    src: path.join(__dirname, 'demo'),
    compile: compile
}));

app.use(
    serveStatic(path.join(__dirname, 'demo'))
);

http.createServer(app).listen(port, function () {
    console.log('Rider demo started on port %d', port);
});


lib/rider.js
/**
 * @file rider
 * @author firede(firede@firede.us)
 */

var helpers = require('./helpers');

/**
 * stylus 插件
 *
 * @param {Object=} options 可选参数
 * @param {boolean=} options.implicit 隐式调用
 * @return {Function}
 */
function plugin(options) {
    options = options || {};
    var implicit = options.implicit;

    if (typeof implicit !== 'boolean') {
        implicit = true;
    }

    return function (style) {
        // 包含 rider 的目录
        style.include(__dirname);

        // 定义通过 JavaScript 扩展的函数
        for (var key in helpers) {
            if (helpers.hasOwnProperty(key)) {
                style.define(key, helpers[key]);
            }
        }

        // 隐式导入 rider
        if (implicit) {
            style.import('rider');
        }
    };
}

module.exports = plugin;

exports.version = require('../package.json').version;


lib/helpers.js
/**
 * @file 帮助函数
 * @author firede(firede@firede.us)
 */

var path = require('path');
var utils = require('stylus').utils;

/**
 * 检查文件是否存在
 *
 * @param {Object} file 文件路径对象
 * @return {boolean}
 */
exports['file-exists'] = function fileExists(file) {
    utils.assertString(file, 'file');
    return !!utils.lookup(file.val, this.paths);
};

/**
 * 检查是否是Url
 * 这里检查的并不严格,主要是为了快速区分path类型
 *
 * @param {Object} url Url路径对象
 * @return {boolean}
 */
exports['is-url'] = function isUrl(url) {
    utils.assertString(url, 'url');

    var obj = url.val.match(/^(https?:)?\/\//i);

    return !!obj;
};

/**
 * 获取添加dppx后缀的图片地址
 *
 * @param {Object} src 图片路径对象
 * @param {Object} dppx dppx对象
 * @param {Object} separator 分隔符对象，默认值为 `@`
 * @return {string}
 */
exports['get-retina-src'] = function getRetinaSrc(src, dppx, separator) {
    utils.assertString(src, 'src');
    utils.assertPresent(dppx, 'dppx');

    // 校验 dppx 的类型
    if (typeof dppx.val !== 'number') {
        throw new Error('TypeError: expected number, but got: ' + dppx);
    }

    var originName = src.val;
    var extName = path.extname(originName);
    var separatorValue = separator ? separator.val : '@';

    return originName.slice(0, -extName.length) + separatorValue + dppx.val + 'x' + extName;
};


demo/server.js
/**
 * @file rider demo
 * @author firede(firede@firede.us)
 */

var express = require('express');
var stylus = require('stylus');
var rider = require('../lib/rider');
var ap = require('autoprefixer-core');
var path = require('path');

var app = express();

function compile(str, filePath) {
    return stylus(str)
        .set('filename', filePath)
        .set('resolve url', true)
        .use(rider())
        .on('end', prefixer);
}

function prefixer(err, css) {
    return ap(
        {
            browsers: ['Android >= 2.3', 'iOS >= 5', 'ExplorerMobile >= 10']
        }
    ).process(css).css;
}

app.set('views', path.join(__dirname, '/views'));
app.set('view engine', 'ejs');
app.use(stylus.middleware({
    src: path.join(__dirname, 'public'),
    compile: compile
}));
app.use(express.static(path.join(__dirname, 'public')));

app.get('/:page?', function (req, res) {
    var page = req.params.page || 'index';
    var viewPath = path.join(app.get('views'), page + '.ejs');
    var fs = require('fs');

    if (fs.existsSync(viewPath)) {
        res.set('Content-Type', 'text/html');
        res.render(page);
    }
    else {
        res.status(404).end();
    }
});

var server = app.listen(8888, function () {
    console.log('Rider example started on port %d', server.address().port);
});
